keeper/src/main/scala/zio/keeper/transport/Connection.scala

```diff
@@ -1,10 +1,197 @@
 package zio.keeper.transport
 
 import zio._
+import zio.clock.Clock
+import zio.duration._
 import zio.stream.ZStream
+import zio.logging.log
+import zio.logging.Logging
+import zio.keeper.ByteCodec
+import zio.keeper.SerializationError
 
-trait Connection[R, E, A] {
-  def send(data: A): ZIO[R, E, Unit]
-  val receive: ZStream[R, E, A]
+trait Connection[-R, +E, -I, +O] { self =>
+  def send(data: I): ZIO[R, E, Unit]
+  val receive: ZStream[R, E, O]
+
+  def batchInput[R1 <: R with Clock with Logging, I1](
+    f: Chunk[I1] => I,
+    maxItems: Int = 32,
+    batchTimeout: Duration = 100.millis,
+    messageBuffer: Int = 256
+  ): ZManaged[R1, Nothing, Connection[R1, E, I1, O]] =
+    batchInputM(is => ZIO.succeedNow(f(is)), maxItems, batchTimeout, messageBuffer)
+
+  def batchInputM[R1 <: R with Clock with Logging, E1 >: E, I1](
+    f: Chunk[I1] => ZIO[R1, E1, I],
+    maxItems: Int = 32,
+    batchTimeout: Duration = 100.millis,
+    messageBuffer: Int = 256
+  ): ZManaged[R1, Nothing, Connection[R1, E1, I1, O]] =
+    ZQueue.bounded[I1](messageBuffer).toManaged(_.shutdown).flatMap { queue =>
+      ZStream
+        .fromQueue(queue)
+        .groupedWithin(maxItems.toLong, batchTimeout)
+        .mapM(
+          l =>
+            f(Chunk.fromIterable(l))
+              .flatMap(self.send)
+              .catchAll(e => log.error("Failed sending batched message.", Cause.fail(e)))
+        )
+        .runDrain
+        .toManaged_
+        .fork
+        .as {
+          new Connection[R1, E1, I1, O] {
+            def send(data: I1): ZIO[R1, E1, Unit] =
+              queue.offer(data).unit
+
+            val receive: ZStream[R, E, O] =
+              self.receive
+
+            val close: UIO[Unit] =
+              self.close
+          }
+        }
+    }
+
+  def biMap[I1, O1](f: I1 => I, g: O => O1): Connection[R, E, I1, O1] =
+    new Connection[R, E, I1, O1] {
+      def send(data: I1): ZIO[R, E, Unit] = self.send(f(data))
+
+      val receive: ZStream[R, E, O1] = self.receive.map(g)
+
+      val close: UIO[Unit] = self.close
```

Comment:
Does it make sense to have `close` here, or should it be pushed to some managed scope outside?

Reply:
Close will be removed

Reply:
close has been removed